src/document_graph/document.cpp

```diff
@@ -0,0 +1,270 @@
+#include <document_graph/document.hpp>
+#include <document_graph/content_group.hpp>
+#include <document_graph/util.hpp>
+#include <eosio/crypto.hpp>
+
+namespace hypha
+{
+
+    Document::~Document() {}
+    Document::Document() {}
+
+    Document::Document(eosio::name contract, eosio::name creator, std::vector<ContentGroup> contentGroups) 
+        : contract{contract}, creator{creator}, content_groups{std::move(contentGroups)}
+    {
+        hash = hashContents();
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, ContentGroup contentGroup)
+    {
+        Document (contract, creator, rollup (contentGroup));
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, Content content)
+    {
+        Document (contract, creator, rollup (content));
+    }
+
+    Document::Document(eosio::name contract, eosio::name creator, const std::string &label, const Content::FlexValue &value)
+    {
+        Document (contract, creator, rollup (Content (label, value)));
+    }
+
+    Document::Document(eosio::name contract, const eosio::checksum256 &_hash) : contract{contract}
+    {
+        document_table d_t(contract, contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(_hash);
+        eosio::check(h_itr != hash_index.end(), "document not found: " + readableHash(_hash));
+
+        id = h_itr->id;
+        creator = h_itr->creator;
+        created_date = h_itr->created_date;
+        certificates = h_itr->certificates;
+        content_groups = h_itr->content_groups;
+        hash = hashContents();
+
+        // this should never happen, only if hash algorithm somehow changed
+        eosio::check (hash == _hash, "fatal error: provided and indexed hash does not match newly generated hash");
+    }
+
+    void Document::emplace () 
+    {
+        require_auth(creator);
+        hash = hashContents();
+
+        document_table d_t(contract, contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(hash);
+
+        // if this content exists already, error out and send back the hash of the existing document
+        eosio::check(h_itr == hash_index.end(), "document exists already: " + readableHash(hash));
+
+        d_t.emplace(contract, [&](auto &d) {
+            id = d_t.available_primary_key(); 
+            created_date = eosio::current_time_point();
+            d = *this;
+        });
+    }
+
+    Document Document::getOrNew (eosio::name _contract, eosio::name _creator, std::vector<ContentGroup> contentGroups) 
```

Comment:
I think it would be a nice to have an overload that takes the hash as input parameter.

Reply:
I've added a constructor that accepts the hash and loads the document.  https://github.com/hypha-dao/document-graph/blob/class-refactor/src/document_graph/document.cpp#L33